### PR TITLE
Parsing Map<Enum,Obj> - 501

### PR DESCRIPTION
**Pull request by TheDoctorOne**
Related to issue #501. 

Parsing "Map<Enum,Obj>" where enum's toString method is overridden and not returning the original name of the constant. 
Test code and backtrace of the issue: https://github.com/google/gson/issues/501#issuecomment-912810928

With the added implementation, it first checks for name, as it was doing before. And if the result does not match with any constant that Enum has, it also checks for the output of the "toString()" method and returns the result of it.

### CODE REVIEW

**Comment by Marcono1234** [comment]
Relates to #1920, which is (partially) about `toString()` being used for map keys during serialization.

**Comment by Chaostical** [comment]
## 
### 
# ****

File: gson/src/main/java/com/google/gson/internal/bind/TypeAdapters.java [hunk 1]
```diff
@@ -803,7 +807,11 @@ public EnumTypeAdapter(final Class<T> classOfT) {
         in.nextNull();
         return null;
       }
-      return nameToConstant.get(in.nextString());
+      String key = in.nextString();
+      T constant = nameToConstant.get(key);
+      if (constant == null)
```
Comment (eamonnmcmanus) [inline gson/src/main/java/com/google/gson/internal/bind/TypeAdapters.java:812]: Google style requires a space between `if` and `(` (which I added) and also braces around the body. Alternatively, you could do this:

```java
return (constant == null) ? stringToConstant.get(key) : constant;
```

**Review by eamonnmcmanus — CHANGES_REQUESTED** [review]
This seems reasonable. Just a small quibble about style.

**Comment by Marcono1234** [comment]
Might be good to add a test for this, for example similar to what you have shown in https://github.com/google/gson/issues/501#issuecomment-912810928.
Possibly also verifying that constant names have higher precedence, e.g.:
```java
enum SwitchedToString {
  A("B"),
  B("A");

  private final String toString;

  SwitchedToString(String toString) {
    this.toString = toString;
  }

  @Override
  public String toString() {
    return toString;
  }
}
```

---

Note that I am not a member of this project so feel free to consider this only as suggestion.

**Review by TheDoctorOne — COMMENTED** [review]
Didn't know that, thx for the heads up

File: gson/src/main/java/com/google/gson/internal/bind/TypeAdapters.java (hunk 1 above)
Comment (eamonnmcmanus) [reply gson/src/main/java/com/google/gson/internal/bind/TypeAdapters.java:812]: Sorry to quibble about this, but can you add braces around the body of the `if`, or rewrite using `?:`?

Comment (TheDoctorOne) [reply gson/src/main/java/com/google/gson/internal/bind/TypeAdapters.java]: Changed to `?:`

**Review by eamonnmcmanus — APPROVED** [review]

**Comment by eamonnmcmanus** [comment]
Thanks! I've run this change against the tests for all of Google's internal code and I didn't find any problems.

**Pull request merged**